lib/avro_turf.rb

```diff
@@ -49,6 +49,7 @@ def encode(data, schema_name: nil, namespace: @namespace)
   # Returns nothing.
   def encode_to_stream(data, schema_name: nil, stream: nil, namespace: @namespace)
     schema = @schema_store.find(schema_name, namespace)
+    Avro::SchemaValidator.validate!(schema, data) if defined?(Avro::SchemaValidator)
```

Comment:
If this is added then validation will be performed twice because the datum writer also calls validation on the datum when writing, and that validation is also using `Avro::SchemaValidator` in cases where that class would be defined.

The difference with the validation performed here is that it will raise a different error. When the new schema validator was added it was done in such a way that the old exceptions were preserved, I assume for back-compatibility.

The way I've handled this in other cases is to rescue the old `Avro::IO::TypeError` and then call `Avro::SchemaValidator.validate!` explicitly to get a better validation error.

Reply:
The point is that the validation done doesn't do as many checks as the SchemaValidator does, especially with the patches I've proposed in the `avro_patches` project (which I'll add to the `avro` project as well later on).

Reply:
With the code that is on avro's master branch:

1.  `Avro::IO::DatumWriter#write_data` calls `Avro::Schema.validate` https://github.com/apache/avro/blob/master/lang/ruby/lib/avro/io.rb#L503
2. `Avro::Schema.validate` calls `Avro::SchemaValidator` https://github.com/apache/avro/blob/master/lang/ruby/lib/avro/schema.rb#L97: 
```ruby
      SchemaValidator.validate!(expected_schema, datum)
```

With the avro-patches gem:

1. `Avro::IO::DatumWriter#write_data` calls `Avro::Schema.valdiate`
https://github.com/salsify/avro-patches/blob/master/lib/avro-patches/logical_types/io.rb#L5
2. `Avro::Schema.validate` calls `Avro::SchemaValidator`
https://github.com/salsify/avro-patches/blob/master/lib/avro-patches/logical_types/schema.rb#L59
```ruby
    Avro::SchemaValidator.validate!(expected_schema, logical_datum, options)
```

Reply:
Hm, interesting... not sure if that changed recently or something because I definitely needed to add this line in a monkey patch in order for validation to happen. Let me tinker a bit more.

Reply:
OK, I figured it out. `write_data` passes `recursive: false` into `validate!`. This will validate only the top level of the schema but not any of the fields inside it.

Reply:
That's correct. `write_data` (in the avro-patches implementation) only validates at the top level because writing happens recursively. So the validation will be performed on the lower levels when they are written. With this change we saw a 5x improvement in encoding performance for some cases: https://github.com/apache/avro/pull/230.

Reply:
Hm... definitely something weird then, because I have cases where this causes failures that don't happen otherwise. Going to dig further again.